File: Backend/src/test/java/lan/dk/podcastserver/service/DatabaseServiceTest.java
package lan.dk.podcastserver.service;

import lan.dk.podcastserver.service.properties.Backup;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.hibernate.search.jpa.FullTextEntityManager;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.mockito.stubbing.Answer;

import javax.persistence.Query;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import static java.util.Objects.nonNull;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.*;

/**
 * Created by kevin on 29/03/2016 for Podcast Server
 */
@Slf4j
@RunWith(MockitoJUnitRunner.class)
public class DatabaseServiceTest {

    @Mock Backup backup;
    @Mock FullTextEntityManager fem;
    @InjectMocks
    DatabaseService databaseService;
    private static final Path NOT_DIRECTORY = Paths.get("/tmp", "foo.bar");
    private Query query;
    private Path backupToCreate;

    @Before
    public void beforeEach() throws IOException {
       Files.deleteIfExists(NOT_DIRECTORY);
       if (nonNull(backupToCreate)) {
            Files.deleteIfExists(backupToCreate);
        }
       //FileSystemUtils.deleteRecursively(Paths.get("/tmp", "foo.bar").toFile());
    }

    @Test
    public void should_reject_if_destination_isnt_directory() throws IOException {
        /* Given */
        Files.createFile(NOT_DIRECTORY);

        /* When */
        Path backupFile = databaseService.backup(NOT_DIRECTORY, true);

        /* Then */
        assertThat(backupFile).isSameAs(NOT_DIRECTORY);
        verify(fem, never()).createNativeQuery(anyString());
    }

    @Test
    public void should_generate_an_archive_of_db_in_binary() throws IOException {
        /* Given */
        when(fem.createNativeQuery(anyString())).then(generateDumpFile());

        /* When */
        Path backupFile = databaseService.backup(Paths.get("/tmp"), true);

        /* Then */
        verify(fem, times(1)).createNativeQuery(contains("BACKUP TO"));
        assertThat(backupFile).exists().hasFileName(backupToCreate.getFileName() + ".tar.gz");
    }

    @Test
    public void should_generate_an_archive_of_db_in_sql() throws IOException {
        /* Given */
        when(fem.createNativeQuery(anyString())).then(generateDumpFile());

        /* When */
        Path backupFile = databaseService.backup(Paths.get("/tmp"), false);

        /* Then */
        verify(fem, times(1)).createNativeQuery(contains("SCRIPT TO"));
        assertThat(backupFile)
                .exists()
                .hasFileName(backupToCreate.getFileName() + ".tar.gz");
    }

    @Test
    public void should_generate_from_backup_parameters() throws IOException {
        /* Given */
        when(fem.createNativeQuery(anyString())).then(generateDumpFile());
        when(backup.getBinary()).thenReturn(Boolean.FALSE);
        when(backup.getLocation()).thenReturn(Paths.get("/tmp"));

        /* When */
        Path backupFile = databaseService.backupWithDefault();

        /* Then */
        verify(fem, times(1)).createNativeQuery(contains("SCRIPT TO"));
        assertThat(backupFile)
                .exists()
                .hasFileName(backupToCreate.getFileName() + ".tar.gz");
    }

    private Answer<Object> generateDumpFile() {
        return i -> {
            backupToCreate = Paths.get(StringUtils.substringAfter(i.getArguments()[0].toString(), "\'").replace("'", ""));
            Files.createFile(backupToCreate);
            query = mock(Query.class);
            return query;
        };
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/service/DatabaseServiceTest.kt
package com.github.davinkevin.podcastserver.service

import com.github.davinkevin.podcastserver.service.properties.Backup
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.contains
import org.mockito.ArgumentMatchers.startsWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.invocation.InvocationOnMock
import org.mockito.junit.jupiter.MockitoExtension
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import javax.persistence.EntityManager
import javax.persistence.Query

/**
 * Created by kevin on 29/03/2016 for Podcast Server
 */
@ExtendWith(MockitoExtension::class)
class DatabaseServiceTest {

    @Mock lateinit var backup: Backup
    @Mock lateinit var em: EntityManager
    @InjectMocks lateinit var databaseService: DatabaseService

    private var query: Query? = null
    private var backupToCreate: Path? = null

    @Nested
    inner class ErrorCases {

        @BeforeEach @AfterEach
        fun beforeEach() {
            Files.deleteIfExists(NOT_DIRECTORY)
        }

        @Test
        fun `should reject if destination is not a directory`() {
            /* Given */
            Files.createFile(NOT_DIRECTORY)

            /* When */
            val backupFile = databaseService.backup(NOT_DIRECTORY, true)

            /* Then */
            assertThat(backupFile).isSameAs(NOT_DIRECTORY)
            verify<EntityManager>(em, never()).createNativeQuery(anyString())
        }
    }

    @Nested
    inner class BackupGeneration {
        @AfterEach
        fun afterEach() {
            Files.deleteIfExists(backupToCreate)
        }

        @Test
        fun `should generate an archive of db in binary format`() {
            /* Given */
            whenever(em.createNativeQuery(anyString())).then { generateDumpFile(it) }

            /* When */
            val backupFile = databaseService.backup(Paths.get("/tmp"), true)

            /* Then */
            verify(em, times(1)).createNativeQuery(contains("BACKUP TO"))
            assertThat(backupFile).exists().hasFileName("${backupToCreate!!.fileName}.tar.gz")
        }

        @Test
        fun `should generate an archive of db in sql format`() {
            /* Given */
            whenever(em.createNativeQuery(startsWith("SCRIPT TO"))).then { generateDumpFile(it) }

            /* When */
            val backupFile = databaseService.backup(Paths.get("/tmp"), false)

            /* Then */
            verify(em, times(1)).createNativeQuery(contains("SCRIPT TO"))
            assertThat(backupFile)
                    .exists()
                    .hasFileName("${backupToCreate!!.fileName}.tar.gz")
        }

        @Test
        fun `should generate from backup parameters`() {
            /* Given */
            whenever(backup.binary).thenReturn(false)
            whenever(backup.location).thenReturn(Paths.get("/tmp"))
            whenever(em.createNativeQuery(startsWith("SCRIPT TO"))).then { generateDumpFile(it) }

            /* When */
            val backupFile = databaseService.backupWithDefault()

            /* Then */
            verify(em, times(1)).createNativeQuery(contains("SCRIPT TO"))
            assertThat(backupFile)
                    .exists()
                    .hasFileName("${backupToCreate!!.fileName}.tar.gz")
        }

        private fun generateDumpFile(i: InvocationOnMock): Query? {
            backupToCreate = Paths.get(i.arguments[0].toString().substringAfter("\'").replace("'", ""))
            Files.createFile(backupToCreate)
            query = mock()
            return query
        }
    }

    companion object {
        private val NOT_DIRECTORY = Paths.get("/tmp", "foo.bar")
    }
}
